### PR TITLE
PoC PR to allow remember keyboard layout for each browser tab (in my case it is Firefox)

### PR DESCRIPTION
**Pull request by dmitry-j-mikhin**
@artemsen, first of all, thank you for this amazing project. I've added some additional functionality to allow you to save and restore your keyboard layout for each browser tab. Some values are hard-coded, such as the LRU cache size and the Firefox name. Perhaps all this functionality should be controlled by an additional runtime flag. But it works for me and I want to share it with you. I think this may be useful in some cases.

### CODE REVIEW

File: src/lrucache.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * File:   lrucache.hpp
+ * Author: Alexander Ponomarev
```
[the remark this answers is not in the document]
Comment (artemsen) [reply src/lrucache.hpp:3]: > It has O(n) complexity for insertion and search

Even if you have 100 windows and 1000 open tabs - O(n) is not a problem.
I don't think anyone would create thousands of windows or open millions of tabs.

> And writing an efficient algorithm in C is very painful

Just an example, hashmap in ~300 lines of C code: http://pokristensson.com/strmap.html

> My perfectionism does not like O(n) complexity even in simple programs

My perfectionism doesn't like mixing C and C++ in a project with 5 source files :)

File: src/sway.h [hunk 2]
```diff
@@ -8,13 +8,13 @@
  * @param[in] window identifier of currently focused window (container)
  * @return keyboard layout to set, -1 to leave current
  */
-typedef int (*on_focus)(int window);
+typedef int (*on_focus)(const char* window_key);
```
Comment (artemsen) [inline src/sway.h:11]: `window_key` that was created by the function `get_cache_key()`.
I think we should use one one name for this: either `const char* cache_key` or `get_window_key()`.

File: src/lrucache.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitry-j-mikhin) [reply src/lrucache.hpp:3]: I understand you) This is just for fun, we are not paid for side projects. I had fun mixing two languages together, in some production system I'll think twice about it) But anyway I just wanted to share the idea. This PR does not have to be merged into your project.

[the remark this answers is not in the document]
Comment (artemsen) [reply src/lrucache.hpp:3]: No problem =)
Thank you anyway.

File: src/sway.h (hunk 2 above)
Comment (dmitry-j-mikhin) [reply src/sway.h:11]: Renamed to get_window_key

File: src/lrucache.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitry-j-mikhin) [reply src/lrucache.hpp:3]: Thanks again for this project! This really saves me a lot of time in Sway)

**Pull request closed without merging**